File: test/RealWorld/search_never_worse.py
#!/usr/bin/env python3
"""FR-50 gate: ``--search`` may never port FEWER items than not searching.

Why this exists as a permanent gate rather than as one more assertion inside
run_realworld.py: the loss it guards against is SILENT. FR-43's search starts
from FR-41's Green-or-Yellow coloring and every child state admits strictly
less than its parent, so an item the coloring wrongly calls Red is absent from
the search's answer and no later stage can put it back. Nothing in the emitted
crate says a thing is missing -- it is simply not there. A coloring change that
reintroduced such a false Red would pass every other test in the tree, because
every other test is written against the coloring's own idea of what should be
in the crate.

So the reference here is deliberately NOT the coloring: it is the unrestricted
recovering import, the one ``--incremental`` performs with nothing excluded.
That import is a fact about the importer, not about the analysis, which is what
makes it a check on the analysis at all.

The measurement is FR-44's own per-item ledger (``emitrust-progress.json``'s
``totals.ported``), so "ported" means here exactly what it means everywhere
else in the project, and both sides are compared with the SAME denominator.

Usage:
  search_never_worse.py --emitrust-cc <path> --corpus <dir> --workdir <dir>
"""

import argparse
import json
import os
import shutil
import subprocess
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from run_realworld import read_compile_commands  # noqa: E402

TIMEOUT_SECONDS = 900


def ported_count(tool, sources, include_flags, crate_dir, search):
    """Transpile once and return (ported, total) from the FR-44 ledger, or None.

    ``search`` selects the compared configuration. Everything else is held
    identical on purpose -- same sources, same include flags, same
    ``--incremental`` -- so the only difference between the two numbers this
    script compares is whether FR-43 ran.
    """
    if os.path.isdir(crate_dir):
        shutil.rmtree(crate_dir)
    argv = [tool, "--emit=crate", "--incremental", *sources, *include_flags,
            "-o", crate_dir]
    if search:
        argv.append("--search")
    try:
        subprocess.run(argv, timeout=TIMEOUT_SECONDS, stdout=subprocess.PIPE,
                       stderr=subprocess.PIPE, check=False)
    except subprocess.TimeoutExpired:
        return None
    try:
        with open(os.path.join(crate_dir, "emitrust-progress.json"),
                  "r", encoding="utf-8") as handle:
            totals = json.load(handle)["totals"]
    except (OSError, ValueError, KeyError):
        return None
    return totals["ported"], totals["graph_items"]


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--emitrust-cc", required=True)
    parser.add_argument("--corpus", required=True)
    parser.add_argument("--workdir", required=True)
    args = parser.parse_args()

    os.makedirs(args.workdir, exist_ok=True)
    projects = sorted(
        name for name in os.listdir(args.corpus)
        if os.path.isfile(os.path.join(args.corpus, name,
                                       "compile_commands.json")))
    if not projects:
        print("FAIL: no corpus project found under %s" % args.corpus)
        return 1

    failures = []
    for name in projects:
        project_dir = os.path.join(args.corpus, name)
        sources, include_dirs = read_compile_commands(project_dir)
        include_flags = ["-I" + d for d in include_dirs]
        plain = ported_count(args.emitrust_cc, sources, include_flags,
                             os.path.join(args.workdir, name + ".plain"),
                             search=False)
        searched = ported_count(args.emitrust_cc, sources, include_flags,
                                os.path.join(args.workdir, name + ".searched"),
                                search=True)
        if plain is None or searched is None:
            failures.append("%s: a configuration produced no readable "
                            "emitrust-progress.json (plain=%s searched=%s)"
                            % (name, plain, searched))
            continue
        verdict = "ok" if searched[0] >= plain[0] else "REGRESSION"
        print("%-14s no-search %d/%d   --search %d/%d   %s"
              % (name, plain[0], plain[1], searched[0], searched[1], verdict))
        if searched[0] < plain[0]:
            failures.append(
                "%s: --search ported %d items, plain --incremental ported %d. "
                "The search can only ever REMOVE items from FR-41's root "
                "state, so this is an item the coloring called Red and the "
                "importer accepts -- a false Red, and it is unrecoverable."
                % (name, searched[0], plain[0]))

    if failures:
        print("")
        for failure in failures:
            print("FAIL: %s" % failure)
        return 1
    print("\nAll %d corpus projects: --search >= no-search." % len(projects))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: test/RealWorld/Cpp/search-never-worse.cpp
// FR-50 anti-regression gate: on every project of the FR-46 C++ corpus,
// `--emit=crate --incremental --search` must port AT LEAST as many items as
// the same command without `--search`.
//
// RUN: %python %S/../search_never_worse.py --emitrust-cc emitrust-cc \
// RUN:   --corpus %S/Inputs --workdir %t | FileCheck %s
//
// This file has a .cpp suffix only so lit discovers it (test/lit.cfg.py sets
// config.suffixes = [".mlir", ".c", ".cpp"]); it contains no C++ code. The
// corpus projects live under Inputs/, which config.excludes keeps out of
// discovery, and are driven through their checked-in compile_commands.json by
// ../search_never_worse.py.
//
// WHY A SEPARATE GATE, next to the RealWorld C++ one that already transpiles
// these same projects. Two reasons, and both are about what the gate is
// checking rather than about what it runs:
//
//  1. It is the only test in the tree whose reference is NOT the coloring.
//     Every FileCheck over `--emit=coloring` or `--emit=search` asserts that
//     the analysis says what it currently says; if FR-41 starts calling a
//     portable item Red, those tests are updated to match and the loss is
//     invisible. Here the reference is the unrestricted recovering import --
//     a fact about the IMPORTER -- so the two sides can disagree, and the
//     disagreement is the failure.
//  2. The failure it guards against is silent and irreversible. FR-43's root
//     state is the Green-or-Yellow set and every child admits strictly fewer
//     items than its parent, so the search can only ever REMOVE. An item the
//     coloring wrongly excludes is gone from the crate with no diagnostic, no
//     `excluded` line a person would question, and no later stage that could
//     recover it.
//
// Unlike realworld-cpp.cpp this needs no `cargo`: nothing is built, only
// transpiled and counted, so the gate runs on every configuration.
//
// The MEASURED numbers as of FR-50 (`fixed-stats` and `tokenizer` are wholly
// inside the subset; `polygon` and `shapes` are mostly outside it and are the
// interesting cases). Both sides are checked, not just the verdict, so a
// change that lowers BOTH counts equally still shows up in the diff.
//
// CHECK: fixed-stats   no-search 11/11   --search 11/11   ok
// Ratcheted 2/8 -> 4/8 by FR-48 (C++ reference parameters), which landed
// after this gate was written. The INVARIANT this file exists to protect is
// the trailing `ok` on every row and the summary line below; the fractions
// are pinned alongside it so a silent DROP is a failure too, and they move
// only forward.
// CHECK: polygon       no-search 4/8   --search 4/8   ok
// CHECK: shapes        no-search 1/7   --search 1/7   ok
// CHECK: tokenizer     no-search 5/5   --search 5/5   ok
// CHECK: All 4 corpus projects: --search >= no-search.
